### PR TITLE
Account uncompressed length while decompressing

### PR DESCRIPTION
**Pull request by buyaa-n**

### CODE REVIEW

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 1]
```diff
@@ -1017,7 +1022,18 @@ public override void Write(byte[] buffer, int offset, int count)
                     int bytesRead = _deflateStream._inflater.Inflate(new Span<byte>(_arrayPoolBuffer));
                     if (bytesRead > 0)
                     {
+                        bool greaterThanUncompressedSize = false;
```
Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: I don't think solely limiting here is sufficient.  I think all the Read members on the stream itself will not go through this codepath, but should also limit.  You could either duplicate the limiting logic there or push it down to the inflater.  Did you consider these options?

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 2]
```diff
@@ -1017,7 +1022,18 @@ public override void Write(byte[] buffer, int offset, int count)
...
+                        {
+                            bytesRead = (int)_deflateStream._uncompressedSize;
+                            greaterThanUncompressedSize = true;
+                        }
+                        else
+                            _deflateStream._uncompressedSize -= bytesRead;
+
                         _destination.Write(_arrayPoolBuffer, 0, bytesRead);
+                        if (greaterThanUncompressedSize)
+                            throw new InvalidDataException(SR.GenericInvalidData);
```
Comment (buyaa-n) [inline src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > I thought we had discussed that there was precedent for truncation since that's what's already happening with compressedSize. 

Yeah we talked about it, but that time i haven't had checked this behavior with other app and after that I feel throwing is better ...
> It's a bit unusual to throw and have a contract that the object is still safe to interact with

The issue reported that throwing is expected, when tried with WinZip, 7zip I got corrupted error messages and only _ucompressedSize section of file was decompressed, well WinZip was asking if I would like to continue unzipping entire file even if it is corrupted. Anyways, that is an option for Apps using this API. But I believe the Apps should know if there were any invalid/corrupted data and choose how to proceed. So preferred to throw, of course open for more discussion

> but I suppose if you document it and test all the edge cases for well defined behavior it might be ok.

All existing tests passed with this change, for now i have only one test for this scenario which is passed, might create more but that leads to another question i wanted to discuss: the existing test files not enough to create more meaningful tests, i tried to add new test file but even uploaded to nuget with new version, some of existing tests failing with new nuget version (some byte difference in Assert), so reverted back that change

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 1 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: If i understood you correctly you mean about WriteAsync(...) method? That was actually my next question to ask from you, like when/how they are called and should i limit their read too, i guess we should limit it too in which case need to have other solution instead of updating uncompressed size with each byteRead

Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: Also there is another deflator DeflateManagedStream need to be covered, which I haven't touched yet

Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > I see, you're modifying the underlying compressed size. I'd avoid that and instead do the math in the check and make the underlying field readonly. Or consider renaming it to "remainingUncompressedBytes". That will make it clearer. I personally prefer the former since it avoids new mutable state on the object.

in case we need to limit async writes, would follow your "former" suggestion

Comment (ericstj) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > you mean about WriteAsync(...) method?

More than that. 

You're trying to make DeflateStream only read and copy up to a certain uncompressed size.

To do that you need to make sure that all Read and Copy methods will either not proceed past the given uncompressed size, or throw.  All of the places I see potentially reading past the compressed size are here:
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L230
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L264
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L398
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L471
https://github.com/dotnet/corefx/blob/7432f96e2af8541bcb2a22cd6a0bd625fe7f71a2/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L939
... probably more.  Looks to me like every time Inflate is called you need to check to make sure it wouldn't exceed the uncompressed size.

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (ericstj) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > The issue reported that throwing is expected

The reported issue is referring to an inconsistency between the local header and central header, for that we should throw.

Did you actually test the scenario where the headers were consistent but the payload was too long?  It's not clear to me if we have sufficient precedent to throw in that case.

> All existing tests passed with this change

That's not what I'm suggesting.  I'm saying that if you document that it is "ok" to handle this exception and still interact with the zip archive and entry, then you really need to test that the state of the objects is still consistent after handling an exception like this.   To test that you'd have to write new tests that catch the thrown exception then fully exercise all relevant functionality.

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 1 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: sure, will check them

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > Did you actually test the scenario where the headers were consistent but the payload was too long? It's not clear to me if we have sufficient precedent to throw in that case.

Yes I did, but for now have only one test for that https://github.com/buyaa-n/corefx/pull/2/files#diff-ee82cb4b115f77168353a3d82a87037cR154. 
By `All existing tests passed with this change` i just wanted to mean this change didn't broke anything

Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > That's not what I'm suggesting. I'm saying that if you document that it is "ok" to handle this exception and still interact with the zip archive and entry, then you really need to test that the state of the objects is still consistent after handling an exception like this. To test that you'd have to write new tests that catch the thrown exception then fully exercise all relevant functionality.

The only test i have only checking the copied stream size after catching the exception, by `then fully exercise all relevant functionality` you mean if the received stream and/or ZAEntry function properly? sure I can add such test

Comment (ericstj) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: I'm suggesting the implications of such an assertion, not necessarily telling you you have to do that.  *If* you say that it's safe to catch that exception and continue using the object then you better make sure all the functionality of the object still works.  I'm trying to illustrate why that's a hard thing to say, since you need to test everything.  These objects are stateful objects, so it can be very tricky to make a claim like its safe to catch an exception and continue using it.  I think you'd be better off either saying you cannot use the object after throw, or just truncate the data (like we truncate the compressed data).

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 1 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: @ericstj I have moved the logic to inflate method and added more tests please re review

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: Sorry still not getting the exact point, but I assume you mean if the read stream will be fully functional, as you know there is 2 read mode, Read only and Update:
1. For read only mode all functionalities should be work
2. For update mode I assume it should work, but yes I need to test that and I will add more test for that

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 3]
```diff
@@ -429,6 +429,80 @@ public static bool TrySkipBlock(BinaryReader reader)
...
         }
+
+        public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
+        {
+            const int OffsetToFilenameLength = 22; // from the point after the signature
+
+            if (reader.ReadUInt32() != SignatureConstant)
+                return false;
+
+            if (reader.BaseStream.Length < reader.BaseStream.Position + OffsetToFilenameLength)
```
Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: You may want to think a bit more about what checks you do for the size.  TrySkipBlock did this check because it was going to seek past it.  You aren't doing an explicit seek of `OffsetToFilenameLength` but instead you are doing fixed-sized reads and then seeks based on the data you've read. I'd also expect comments that indicate what information you are seeking past/seeking-to.

You're missing the check on fileNameLength / extra-FieldLength.

As a whole, try to think about garunteeing you have enough data available to proceed and when you read an offset that isn't fixed size make sure you check up front that you can seek to that offset.

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 4]
```diff
@@ -429,6 +429,80 @@ public static bool TrySkipBlock(BinaryReader reader)
...
+                return false;
+
+            uint minimunVersion = reader.ReadUInt16();
+            int dataDescriptorBit = reader.ReadUInt16() & 0x08;
```
Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: This should use https://github.com/dotnet/corefx/blob/0eed4042f056b4d9f6f4d1bf2918e234d67a6f88/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L1253.

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 5]
```diff
@@ -429,6 +429,80 @@ public static bool TrySkipBlock(BinaryReader reader)
...
+            long compressedSize = reader.ReadUInt32();
+            long uncompressedSize = reader.ReadUInt32();
+            int filenameLength = reader.ReadUInt16();
+            uint extraFieldLength = reader.ReadUInt16();
+            reader.BaseStream.Seek(filenameLength, SeekOrigin.Current);
+            long endExtraFields = reader.BaseStream.Position + extraFieldLength;
+            
+            if (dataDescriptorBit == 0)
+            {
+                bool uncompressedSizeInZip64 = uncompressedSize == ZipHelper.Mask32Bit;
```
Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: nit: consider prefixing these with `is` to make it clearer that they are bools and not sizes.

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 6]
```diff
@@ -429,6 +429,80 @@ public static bool TrySkipBlock(BinaryReader reader)
...
+        {
+            const int OffsetToFilenameLength = 22; // from the point after the signature
+
+            if (reader.ReadUInt32() != SignatureConstant)
+                return false;
+
+            if (reader.BaseStream.Length < reader.BaseStream.Position + OffsetToFilenameLength)
+                return false;
+
+            uint minimunVersion = reader.ReadUInt16();
```
Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: nit: spelling

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 7]
```diff
@@ -429,6 +429,80 @@ public static bool TrySkipBlock(BinaryReader reader)
...
+                        compressedSize = zip64.CompressedSize.Value;
+                }
+
+                reader.BaseStream.AdvanceToPosition(endExtraFields);
+            }
+            else
+            {
+                bool is64bit = minimunVersion > 44;
+                long seekSize = entry.CompressedLength + (is64bit ? 8 : 4) * 2;
+                uint dataDescriptorSignature = reader.ReadUInt32();
```
Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: I think you're reading this from the wrong location.  At this point the stream is still pointing in the local header, but the data descriptor signature should be at the end of the file.  I believe you end up skiping over the actual data descriptor signature on line 484 below (when you added + 4).

Comment (ericstj) [inline src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: I don't believe that this is the correct place to read an optional descriptor signature.  If I look at our code that writes it, it's writing immediately before the CRC in the descriptor itself.  So I think you need to change the code at 484 to account for the optional descriptor there preceeding the CRC.  See https://github.com/dotnet/corefx/blob/0eed4042f056b4d9f6f4d1bf2918e234d67a6f88/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L1023-L1045.

**Comment by ericstj** [comment]
I think you can move this over to corefx now.  I believe you have the primary thrust of the changes that need to be made.  The rest of the work will likely be responding to feedback and you'd benefit from getting that from more folks than just me.

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 3 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Sure , will review all again and add needed check

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (ericstj) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: It's more like: if you tell someone it is OK to catch an exception and still use the stream, are you garunteeing that the stream is in a safe state?  Here's a list of the sort of things I'd expect to work:
1. Position represents a legitimate value and not some value greater than the size you were trying to limit.
2. Subsequent reads either return nothing or will throw in the same way.
3. Seeks on the stream do a reasonable thing (either throw, or behave as if seeking from the limited value).
4. many others

Essentially, you need to be *very careful* that the state the stream remains in provides legitimate behavior if you're telling folks its safe to catch an exception.

I strongly believe that you're better off implmenting this as truncation first, so that you get all the behaviors correct.  Then you can decide if you want to throw when it truncates.

File: src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 7 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Good catch, thanks

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > Essentially, you need to be very careful that the state the stream remains in provides legitimate behavior if you're telling folks its safe to catch an exception.
I strongly believe that you're better off implementing this as truncation first, so that you get all the behaviors correct. Then you can decide if you want to throw when it truncates.

Yes truncating the output stream fist and throwing after that, as deflate stream is closed in finally block base deflate stream is not readable anymore. 

> Here's a list of the sort of things I'd expect to work:
> 1. Position represents a legitimate value and not some value greater than the size you were trying to limit.
> 2. Subsequent reads either return nothing or will throw in the same way.
> 3. Seeks on the stream do a reasonable thing (either throw, or behave as if seeking from the limited value).
> 4. many others

Added assertions for base deflate stream, will add more test for output read stream is working as expected. Also will add tests for update mode

**Comment by buyaa-n** [comment]
Sure I will, after adding some more tests

File: src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: Have got issue with update mode. in this mode when `InvalidDataException` occurs it nulls output stream https://github.com/buyaa-n/corefx/pull/2/files#diff-ba8d5a9ef0ce8bd24cf0576f0ab0972aL367
Now i don't know what to do:

1.  allow reading stream after throw only for `Read` mode and leave Update mode as is. 
2. or don't throw and return only truncated stream, in that case i assume base compressed stream remain open/readable which will be no use of this fix
3. or just throw and don't truncate read value, in that case as inflate and read/write operation is done in chunks user will still received truncated stream up to before throw
4. or maybe throw any other exception than `InvalidDataException` and let both modes use truncated stream, in that case Update mode need to be tested thoroughly

This is blocking me to raise PR in corefx, or maybe raise PR hear others opinion?

Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: After thinking about this little more i think the solution number 1 or 4 is more meaningful and what i have on this PR is solution 1, so am going to raise PR on coreFX and see others feedback

Comment (ericstj) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: Having Read throw and Update only truncate seems reasonable.  Not sure if that's what you meant by 1.  You could implement with an additional optional property passed to the inflater, just catch the exception, or limit the bytes copied out to the memory stream during the read portion of Update.

Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: Option 1 is what I have now, throw `InvalidDataException` and return truncated stream, in case of Read mode stream will be readable up to uncompressed size, but in case of Update mode it catching the exception and setting returning truncated steam to null. With that i mean we throw for both case but  in update mode returning stream will not be readable.

The solution you are suggesting is closer to option 4. Where we could throw with something else other than `InvalidDataException` so that in update mode the returning stream could be readable. We need to throw both cases to close the underlying deflate stream and make it not readable more. Actually there is no separate read for Update mode, as it reads the compressed stream using read mode and wraps the returned stream with seekable one

Comment (ericstj) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: I understand that but you can always make the behavior different if you want since you are creating the streams and are able to pass state in.

I really think that you can solve all of these issues if you just truncate the data.  The library is **already** truncating if compressed size is smaller than the compressed payload, the fact that you are throwing when the uncompressed size is smaller than compressed size, and still trying to make the stream be in a usable state after catching the exception is inconsistent with existing behavior and proving to be difficult.

Comment (buyaa-n) [reply src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: > I really think that you can solve all of these issues if you just truncate the data

I didn't found a way to stop decompress without throwing, but I have got one idea might work, will give it try.

>  the fact that you are throwing when the uncompressed size is smaller than compressed size

just a little correction: throwing when decompressed amount goes over uncomressed size
